### PR TITLE
Benchmarks

### PR DESCRIPTION
**Pull request by yosiat**
- Run benchmarks on CI using https://github.com/benchmark-action/github-action-benchmark
- Adding CityLots benchmark

### CODE REVIEW

**Comment by yosiat** [comment]
@embano1 Hi!

I am trying to setup the benchmarks action, I introduced some slow down in CityLots to get a regression, and I am getting this error of `Error: Resource not accessible by integration` when it try posting to Github.

https://github.com/timbray/quamina/actions/runs/3232693534/jobs/5293615852

I added permissions of pull-request to the job as suggested here (https://github.com/actions/first-interaction/issues/10#issuecomment-1232740076) , Do you have any idea what's going here?

Maybe the Github Token don't have permissions for doing writes in PRs?

**Comment by embano1** [comment]
> Maybe the Github Token don't have permissions for doing writes in PRs?

Yeah, this seems to be the common `pull_request` from `fork` issue. IIRC, the `token` won't get elevated privileges here, even though you specify. Can you run this with the comment option set to `false`? Perhaps that's ok for our use case?

**Comment by timbray** [comment]
Sorry Yosi, I'm maxed out on another deliverable, I'll get to this PR by end-of-day or latest tomorrow morning.

**Comment by yosiat** [comment]
@embano1 Gotcha, I'll disable the comment and just make sure the benchmarks are failing properly

@timbray That's ok, it's not yet ready for review ~!

**Comment by yosiat** [comment]
@embano1 

I changed it so CI will fail if there is a regression, but commenting on Pull Request I think is going to be helpful, the results are hard to read (https://github.com/timbray/quamina/actions/runs/3350891275/jobs/5551990064) while in Github comments there easier to make-sense.

> IMHO it's not worth opening the security door here

Can you elaborate on the security issue you see here? because currently we have "Codecov" doing comments on Github Pull Requests, and we are ok with that, so it's hard for me to see difference between codecov and benchmarks actions.

**Comment by timbray** [comment]
Let me know when you want a close look. Would like to get this merged.

**Comment by yosiat** [comment]
@timbray it's actually ready for review, the only problem for me is that we can't comment on pr, since it's blocking me, I'll disable it for now.

**Comment by embano1** [comment]
@yosiat sorry for my delay and slight confusion about the error message you reported earlier.

I just checked the `action` in use again and the `Error: Resource not accessible by integration` you got. It seems this is thrown not because the `action` doesn't work with comments in PRs (it does of course), but because you ran this check against the upstream branch which your token does not have access to. 

Did you verify the `action` works in one of your repos? We have two options here:

1) merge as is and see if it works :)
2) you perform some additional testing in a repo you own, e.g. your fork but open the PR against your fork and report back here

Thoughts?

File: .github/workflows/benchmarks.yml
Comment (embano1) [inline .github/workflows/benchmarks.yml:35]: What is the default of how long the cache content is retained? E.g. if we don't have PRs and updated benchmark data pushed to the cache, do we lose the data at some point and then miss regressions for some reasons?

File: .github/workflows/benchmarks.yml
Comment (embano1) [inline .github/workflows/benchmarks.yml:41]: Can we pin the action with a the SHA syntax used above?

File: citylots_bench_test.go
Comment (embano1) [inline citylots_bench_test.go:7]: IMHO the benchmark looks good to avoid compiler optimizations, but just to double check did you run with `go test -gcflags "-m -m" -bench=<...>` and check for inlining?

Comment (yosiat) [reply citylots_bench_test.go:7]: Didn't double checked, but now I checked and I don't see inlining here -

```
./citylots_bench_test.go:48:42: err.Error() escapes to heap:
./citylots_bench_test.go:48:42:   flow: {storage for ... argument} = &{storage for err.Error()}:
./citylots_bench_test.go:48:42:     from err.Error() (spill) at ./citylots_bench_test.go:48:42
./citylots_bench_test.go:48:42:     from ... argument (slice-literal-element) at ./citylots_bench_test.go:48:12
./citylots_bench_test.go:48:42:   flow: {heap} = {storage for ... argument}:
./citylots_bench_test.go:48:42:     from ... argument (spill) at ./citylots_bench_test.go:48:12
./citylots_bench_test.go:48:42:     from (*testing.common).Errorf(b.common, "Matches4JSON: %s", ... argument...) (call parameter) at ./citylots_bench_test.go:48:12
./citylots_bench_test.go:34:20: matcherStats(q.matcher.(*coreMatcher)) escapes to heap:
./citylots_bench_test.go:34:20:   flow: {storage for ... argument} = &{storage for matcherStats(q.matcher.(*coreMatcher))}:
./citylots_bench_test.go:34:20:     from matcherStats(q.matcher.(*coreMatcher)) (spill) at ./citylots_bench_test.go:34:20
./citylots_bench_test.go:34:20:     from ... argument (slice-literal-element) at ./citylots_bench_test.go:34:7
./citylots_bench_test.go:34:20:   flow: {heap} = {storage for ... argument}:
./citylots_bench_test.go:34:20:     from ... argument (spill) at ./citylots_bench_test.go:34:7
./citylots_bench_test.go:34:20:     from (*testing.common).Log(b.common, ... argument...) (call parameter) at ./citylots_bench_test.go:34:7
./citylots_bench_test.go:31:47: err.Error() escapes to heap:
./citylots_bench_test.go:31:47:   flow: {storage for ... argument} = &{storage for err.Error()}:
./citylots_bench_test.go:31:47:     from err.Error() (spill) at ./citylots_bench_test.go:31:47
./citylots_bench_test.go:31:47:     from ... argument (slice-literal-element) at ./citylots_bench_test.go:31:12
./citylots_bench_test.go:31:47:   flow: {heap} = {storage for ... argument}:
./citylots_bench_test.go:31:47:     from ... argument (spill) at ./citylots_bench_test.go:31:12
./citylots_bench_test.go:31:47:     from (*testing.common).Fatalf(b.common, "AddPattern failed: %s", ... argument...) (call parameter) at ./citylots_bench_test.go:31:12
./citylots_bench_test.go:29:27: names[i] escapes to heap:
./citylots_bench_test.go:29:27:   flow: x = &{storage for names[i]}:
./citylots_bench_test.go:29:27:     from names[i] (spill) at ./citylots_bench_test.go:29:27
./citylots_bench_test.go:29:27:     from q, x, patternJSON := q, names[i], patterns[i] (assign-pair) at ./citylots_bench_test.go:29:21
./citylots_bench_test.go:29:27:   flow: {heap} = x:
./citylots_bench_test.go:29:27:     from q.matcher.addPattern(x, patternJSON) (call parameter) at ./citylots_bench_test.go:29:21
./citylots_bench_test.go:26:34: err.Error() escapes to heap:
./citylots_bench_test.go:26:34:   flow: {storage for ... argument} = &{storage for err.Error()}:
./citylots_bench_test.go:26:34:     from err.Error() (spill) at ./citylots_bench_test.go:26:34
./citylots_bench_test.go:26:34:     from ... argument (slice-literal-element) at ./citylots_bench_test.go:26:11
./citylots_bench_test.go:26:34:   flow: {heap} = {storage for ... argument}:
./citylots_bench_test.go:26:34:     from ... argument (spill) at ./citylots_bench_test.go:26:11
./citylots_bench_test.go:26:34:     from (*testing.common).Fatalf(b.common, "New(): %s", ... argument...) (call parameter) at ./citylots_bench_test.go:26:11
./citylots_bench_test.go:7:24: parameter b leaks to {heap} with derefs=0:
./citylots_bench_test.go:7:24:   flow: {heap} = b:
./citylots_bench_test.go:7:24:     from b.common (dot of pointer) at ./citylots_bench_test.go:26:4
./citylots_bench_test.go:7:24:     from b.common (address-of) at ./citylots_bench_test.go:26:4
./citylots_bench_test.go:7:24:     from (*testing.common).Fatalf(b.common, "New(): %s", ... argument...) (call parameter) at ./citylots_bench_test.go:26:11
./citylots_bench_test.go:7:24: leaking param: b
./citylots_bench_test.go:10:22: []string{...} does not escape
./citylots_bench_test.go:16:19: []string{...} does not escape
./citylots_bench_test.go:26:11: ... argument does not escape
./citylots_bench_test.go:26:34: err.Error() escapes to heap
./citylots_bench_test.go:29:27: names[i] escapes to heap
./citylots_bench_test.go:31:12: ... argument does not escape
./citylots_bench_test.go:31:47: err.Error() escapes to heap
./citylots_bench_test.go:34:7: ... argument does not escape
./citylots_bench_test.go:34:20: matcherStats(q.matcher.(*coreMatcher)) escapes to heap
./citylots_bench_test.go:48:12: ... argument does not escape
./citylots_bench_test.go:48:42: err.Error() escapes to heap
```

Comment (timbray) [reply citylots_bench_test.go:7]: Why would we want to avoid compiler optimizations?  The benchmark should give a realistic read-out of what performance people should expect?

**Comment by yosiat** [comment]
@embano1 

> I just checked the action in use again and the Error: Resource not accessible by integration you got. It seems this is thrown not because the action doesn't work with comments in PRs (it does of course), but because you ran this check against the upstream branch which your token does not have access to.

If I understand correctly, it happens because I did it from a fork branch which uses *my* token which don't have access to do a comment, if that's the case - this not useful, since we want to help external contributors as well.

I tested it on a repo, and did a PR which slows down everything: https://github.com/yosiat/go-bench-test/commit/33faa612174fd240281afe267edacdd262d893e6 and got a nice comment.

Anyway, CI action that's failing is a good start.

File: citylots_bench_test.go
Comment (yosiat) [reply citylots_bench_test.go:7]: @timbray you want to avoid compiler optimisations in the benchmark it self, so the compiler for example won't eliminate the benchmark loop, for example:
```
func BenchmarkFib20(b *testing.B) {
	for i := 0; i < b.N; i++ {
		Fib(20)
	}
}
```

In this case, the compiler can eliminate the calls to "Fib(20)", so you want try eliminate those issues (but not in the code under benchmark, in our case the matcher/flattener)

**Review by timbray — APPROVED** [review]
I'm just learning the CI and benchmarking stuff so I don't fully understand the issues here, but whatever you guys work out is OK by me.  Might be best to merge so make progress easier?

**Comment by yosiat** [comment]
@timbray thanks, we will merge once I get @embano1 will approve, he knows way better than me on CI and asking me good questions here (like cache lifecycle).

**Comment by timbray** [comment]
+1

File: .github/workflows/benchmarks.yml
Comment (yosiat) [reply .github/workflows/benchmarks.yml:35]: According to github, the cache will be maintained for 7 days if it wasn't accessed. I am not sure what will be the behaviour if we don't touch quamina's main branch for more than 7 days.

https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#usage-limits-and-eviction-policy

File: .github/workflows/benchmarks.yml
Comment (yosiat) [reply .github/workflows/benchmarks.yml]: Changed.

**Comment by embano1** [comment]
> If I understand correctly, it happens because I did it from a fork branch which uses *my* token which don't have access to do a comment, if that's the case - this not useful, since we want to help external contributors as well.

Yes, correct. Just reread my initial response and the reason I said "not open door" was that the only way to fix the comment issue is `pull_request_target` which is a BIG no (https://securitylab.github.com/research/github-actions-preventing-pwn-requests/). 

So we'll have to live without nice comments I'm afraid.

File: citylots_bench_test.go
Comment (embano1) [reply citylots_bench_test.go:7]: Exactly @yosiat ! Thx for confirming.

**Review by embano1 — APPROVED** [review]
LGTM

**Pull request merged**
